src/Symfony/Component/DependencyInjection/Resource/ServicesTrait.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Resource;
+
+trait ServicesTrait
+{
+    private array $services = [];
+
+    /**
+     * @var array<array{
+     *     id: string,
+     *     class?: string|int|float|bool,
+     * }>|array<string, class-string|null> $services
```

Comment:
This is minimalistic. In reality, we could be way more descriptive here, see what YamlFileLoader describes for services. Maybe you preferred postponing this to a follow up?

Reply:
I started implementing what's left (lazy, abstract, args, synthetic, etc.) but there's a bit more work than I expected. I propose to validate as-is, then comeback to this bit when the rest is OK?